pkg/scheduler/schedule_one.go

```diff
@@ -150,8 +151,9 @@ func (sched *Scheduler) schedulingCycle(ctx context.Context, state *framework.Cy
 			nominatingInfo = clearNominatedNode
 			klog.ErrorS(err, "Error selecting node for pod", "pod", klog.KObj(pod))
 			metrics.PodScheduleError(fwk.ProfileName(), metrics.SinceInSeconds(start))
+			reason = SchedulerError
 		}
-		sched.FailureHandler(ctx, fwk, podInfo, err, v1.PodReasonUnschedulable, nominatingInfo)
+		sched.FailureHandler(ctx, fwk, podInfo, err, reason, nominatingInfo)
```

Comment:
This is a break change, since we'll change the reason when updating pod's status. I have no idea whether this was intended, but IMO, I think we should keep the metrics and reason in consistent. cc @Huang-Wei @alculquicondor @ahg-g

Reply:
Do we consider `reason` values part of the API after https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures?

Does CA rely on this value? I am guessing changing this to a different value is not a an issue with CA since we shouldn't be scaling up because of a scheduling error (and CA will anyway not scale up if it indeed fits)

Reply:
I _guess_ it wouldn't break CA. Its logic to find unschedulabe pods is like: `spec.nodeName==,status.phase!=Succeeded,status.phase!=Failed`:

https://github.com/kubernetes/autoscaler/blob/81d70f94ad2094b2a20f1c8e52d15aff8df7bac1/cluster-autoscaler/utils/kubernetes/listers.go#L183-L194

However, in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-up-work, it says:

> Whenever a Kubernetes scheduler fails to find a place to run a pod, it sets "schedulable" PodCondition to false and reason to "unschedulable".

So cc @x13n for confirmation. (TL;DR, we want to figure out if scheduler updates a pod with an internal ScheduleError reason, would it break CA or not?)

Reply:
> Do we consider `reason` values part of the API after https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures?

From the description https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#using-pod-statusreason-field, `reason` is not considered as it's arbitrary and hard to track between different versions. cc @alculquicondor @mimowo for confirmation.

Reply:
If the reason is in the v1 package, it's considered part of the API. We don't filter based on it for Job failure policy, but that's a specific feature. In any case, we should be good to add new values.

@x13n an error is expected in the cases where apiserver is unavailable (this is already the case), or there are parsing errors #111791 (here we are currently returning Unschedulable). So yes, in general CA should skip these pods. This means that CA is not skipping these errors, which is actually a bug.